docs/core-concepts/recording-gameplay.mdx

```diff
@@ -0,0 +1,283 @@
+    <figcaption>
+        Looping the recording displays a counter for the current loop iteration.
+        Looping works best when the recording returns itself to a known start state
+        to set up the next loop iteration.
+    </figcaption>
+</figure>
+</div>
+
+</div>
+
+#### Programmatic Replay
+
+The `RGTestUtils` class provides methods for replaying recordings programmatically.
+This is intended for use in
+[Play Mode tests for the Unity Test Runner](https://docs.unity3d.com/Packages/com.unity.test-framework@1.4/manual/index.html).
+
+:::info
+The `RegressionGames` and `RegressionGames.TestFramework` assemblies must be listed as
+references in your test assembly to use the `RGTestUtils` class.
+
+<img src={TestAssemblyRefs} alt="Test assembly references for Regression Games SDK" style={{display: "block", margin: "auto"}}/>
+:::
+
+:::tip
+By default, [recordings are saved outside your project](#recorded-data-formats).
+Make sure you copy the recordings you'd like to replay into your project
+to ensure they'll be available on other machines that run these automated tests.
+:::
+
+Create a `UnityTest` that invokes the `RGTestUtils.StartPlaybackFromZipFile` method.
+Pass the path to your recording's `bot_segments.zip` file.
+This will replay the recording, wait until playback is complete, and then save the replay as a new recording.
+You can add additional logic to your test to validate the game state at the end of playback,
+or use a recording to reach a known state in your game for further testing.
+
+```csharp
+using System.Collections;
+using RegressionGames;
+using RegressionGames.Types;
+using UnityEngine.Assertions;
+using UnityEngine.SceneManagement;
+using UnityEngine.TestTools;
+
+public class RecordingTest
+{
+
+    [UnityTest]
+    public IEnumerator TestRecording()
+    {
+
+        // Unity has a macOS bug that prints errors when taking screenshots during recording.
+        // This line prevents the test from failing when these errors are logged.
+#if UNITY_EDITOR_OSX || UNITY_RUNTIME_OSX
+        LogAssert.ignoreFailingMessages = true;
+#endif
+
+        // Define which recording to use
+        string recordingPath = System.IO.Path.Combine(System.IO.Directory.GetCurrentDirectory(), "Assets/Tests/Runtime/Recordings/MainMenuRecording.zip");
+
+        // Wait for the scene to load
+        SceneManager.LoadSceneAsync("Startup", LoadSceneMode.Single);
+        yield return RGTestUtils.WaitForScene("MainMenu");
+
+        // Start playback
+        // This will automatically create a new recording of the replay
+        PlaybackResult playbackResult = null;
+        yield return RGTestUtils.StartPlaybackFromZipFile(recordingPath, result => playbackResult = result);
+
+        // Print out the recording path for viewing later
+        RGDebug.LogInfo("Played back and recorded the test run - saved to " + playbackResult.saveLocation);
+        Assert.IsNotNull(playbackResult.saveLocation);
```

Comment:
This was a bug in my code - these should be swapped, and we need to add another assert. There are situations where playbackResult is null.

```suggestion
        Assert.IsNotNull(playbackResult);
        Assert.IsNotNull(playbackResult.saveLocation);
        RGDebug.LogInfo("Played back and recorded the test run - saved to " + playbackResult.saveLocation);

```

Reply:
Accepted your suggestion, but can you elaborate on what situations cause the playbackResult to be null? Is this when playback encounters some error?

Reply:
Exactly, this happens when the playback fails... so this is kind of actually a bug in our own system, but I haven't had time to dig deeper into this yet.